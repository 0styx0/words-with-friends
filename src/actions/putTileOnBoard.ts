import types from './types';
import Tile from '../interfaces/Tile';
import Player from '../classes/Player';

export default function putTileOnBoard(
    tile: Tile, coordinates: ReadonlyArray<number>, currentPlayer: Readonly<Player>, currentTurn: number
) {
    return {
        type: types.PLACE_TILE_ON_BOARD,
        tile,
        coordinates,
        currentPlayer,
        currentTurn
    };
}