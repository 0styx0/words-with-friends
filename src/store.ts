import reducers from './reducers';
import { createStore, compose } from 'redux';
import Player from './classes/Player';
import Tilebag from './classes/Tilebag';
import Board from './classes/Board';

const enhancers = compose('devToolsExtension' in window ?
    (window as {} as { devToolsExtension: Function }).devToolsExtension!() :
    (f: {}) => f);

export const defaultState = {
    turn: 0,
    Players: [new Player(true, 0), new Player(false, 1)], // turns switched around on INIT_PLAYERS
    board: new Board(),
    Tilebag: new Tilebag()
};

const store = createStore(reducers, defaultState, enhancers);

export default store;
