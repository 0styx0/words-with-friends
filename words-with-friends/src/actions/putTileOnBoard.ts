import types from './types';
import Tile from '../interfaces/Tile';

export default function putTileOnBoard(tile: Tile, coordinates: string) {
    return {
        type: types.PLACE_TILE,
        tile,
        coordinates
    };
}