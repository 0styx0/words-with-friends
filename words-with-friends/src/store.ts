import reducers from './reducers';
import { createStore } from 'redux';

export const defaultState = {
    turn: 0
};

const store = createStore(reducers, defaultState);

export default store;
