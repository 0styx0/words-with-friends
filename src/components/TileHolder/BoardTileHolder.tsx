import * as React from 'react';
import TileHolder from './';
import TileType from '../../interfaces/Tile';

import { bindActionCreators } from 'redux';
import { connect, Dispatch } from 'react-redux';
import actionCreators from '../../actions';
import { defaultState } from '../../store';
import Player from '../../classes/Player';
import AbstractTileHolder from './AbstractTileHolder';
import Board from '../../classes/Board';


function mapStateToProps(state: typeof defaultState, props: Props) {

    return {
        coordinates: props.coordinates,
        tile: state.board.get(props.coordinates) ? state.board.get(props.coordinates)!.tile : undefined,
        currentPlayer: state.Players.find(player => player.turn),
        board: state.board,
        turn: state.turn
    };
}

function mapDispatchToProps(dispatch: Dispatch<typeof defaultState>) {
    return bindActionCreators(actionCreators, dispatch);
}


type Props = typeof actionCreators & typeof defaultState & {
    tile?: TileType;
    coordinates: number[];
    currentPlayer: Player;
    board: Board,
    turn: number
};

export class BoardTileHolderContainer extends AbstractTileHolder<Props> {

    /**
     * Puts tile down (@see tileTarget.drop)
     */
    putTile(tile: TileType) {

        this.props.putTileOnBoard(tile, this.props.coordinates, this.props.Players, this.props.turn);
    }

    /**
     * Removes tile from TileHolder (@see Tile.tsx#source.endDrag)
     */
    removeTile() {

        this.props.removeTileFromBoard(this.props.coordinates);
    }

    render() {

        return (
            <div
              className="tileHolder"
              onDrop={this.onDrop}
              onDragOver={this.onDragOver}
            >
                <TileHolder
                    coordinates={this.props.coordinates!}
                    tile={this.props.tile}
                    powerup={this.props.board.get(this.props.coordinates) &&
                        this.props.board.get(this.props.coordinates)!.powerup}
                    removeTile={this.removeTile}
                />
            </div>
        );
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(BoardTileHolderContainer as any);
