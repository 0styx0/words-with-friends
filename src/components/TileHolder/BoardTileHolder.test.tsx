import * as React from 'react';
import * as renderer from 'react-test-renderer';
import store, { getState } from '../../store';
import { Provider } from 'react-redux';
import BoardTileHolder, { BoardTileHolderContainer } from './BoardTileHolder';
import mockMath from '../../test/mocks/Math';
import { mount } from 'enzyme';
import putTileOnBoard from '../../actions/putTileOnBoard';
mockMath();

describe('<BoardTileHolder />', () => {

    describe('renders', () => {

        function snap(playerIndex: number) {

            const tree = renderer.create(
                <Provider store={store}>
                    <BoardTileHolder coordinates={[0, 0]} {...[] as any} />
                </Provider>
            ).toJSON();

            expect(tree).toMatchSnapshot();
        }

        it(`correctly for current player's tile`, () => snap(0));

        it(`correctly for other player's tile`, () => snap(1));
    });

    function setup() {

        const coordinates = [0, 0];
        const wrapper = mount(

            <Provider store={store}>
                <BoardTileHolder Players={getState().Players} coordinates={coordinates} {...[] as any} />
            </Provider>
        );

        return {
            wrapper,
            component: wrapper.find(BoardTileHolderContainer).instance() as BoardTileHolderContainer,
            coordinates
        };
    }

    describe('#putTile', () => {

        it('adds tile to board', () => {

            const { component, coordinates } = setup();

            const currentCoordinateTile = getState().board.get(coordinates);

            expect(currentCoordinateTile).toBeFalsy();

            const newTile = getState().Tilebag.getRandomTile(0);

            component.putTile(newTile);

            const newCoordinateTile = getState().board.get(coordinates)!.tile;

            expect(newCoordinateTile).toEqual(newTile);
        });
    });

    describe('#removeTile', () => {

        it('removes tile as a prop', () => {

            const { component, coordinates } = setup();

            const initialTile = { points: -1, letter: 'invalid letter', playerIndex: 0 };

            const state = getState();
            store.dispatch(
                putTileOnBoard(initialTile, coordinates, state.Players, state.turn)
            );

            const currentCoordinateTile = getState().board.get(coordinates);

            expect(currentCoordinateTile).toBeTruthy();

            component.removeTile();

            const newCoordinateTile = getState().board.get(coordinates)!.tile;

            expect(newCoordinateTile).toBeFalsy();
        });
    });
});

