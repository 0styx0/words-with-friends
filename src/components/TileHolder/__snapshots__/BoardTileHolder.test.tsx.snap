// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BoardTileHolder /> renders correctly for current player's tile 1`] = `
<div
  className="tileHolder"
  onDragOver={[Function]}
  onDrop={[Function]}
>
  <span>
    
  </span>
</div>
`;

exports[`<BoardTileHolder /> renders correctly for other player's tile 1`] = `
<div
  className="tileHolder"
  onDragOver={[Function]}
  onDrop={[Function]}
>
  <span>
    
  </span>
</div>
`;
